editor/plugins/visual_shader_editor_plugin.cpp

```diff
@@ -1227,8 +1227,88 @@ void VisualShaderEditor::_update_options_menu() {
 	Vector<AddOption> custom_options;
 	Vector<AddOption> embedded_options;
 
+	static Vector<String> type_filter_exceptions;
+	if (type_filter_exceptions.is_empty()) {
+		type_filter_exceptions.append("VisualShaderNodeExpression");
+	}
+
 	for (int i = 0; i < add_options.size(); i++) {
 		if (!use_filter || add_options[i].name.findn(filter) != -1) {
+			// port type filtering
+			if (members_output_port_type != VisualShaderNode::PORT_TYPE_MAX || members_input_port_type != VisualShaderNode::PORT_TYPE_MAX) {
+				Ref<VisualShaderNode> vsn;
+				int check_result = 0;
+
+				if (!add_options[i].is_custom) {
+					vsn = Ref<VisualShaderNode>(Object::cast_to<VisualShaderNode>(ClassDB::instantiate(add_options[i].type)));
```

Comment:
The `Ref<>()` constructor might not be needed here, assigning the result of the `Object::cast_to()` should work fine, no?

Reply:
Unfortunately no - due to ambiguity of operators.

Reply:
I mean, why is this situation here:
```
Ref<VisualShaderNode> vsn;
vsn = Ref<VisualShaderNode>(Object::cast_to<VisualShaderNode>(ClassDB::instantiate(add_options[i].type)));
```
different from other casts below?
```
Ref<VisualShaderNodeInput> input = Object::cast_to<VisualShaderNodeInput>(vsn.ptr());
```

Wouldn't this work the same?
```
Ref<VisualShaderNode> vsn;
vsn = Object::cast_to<VisualShaderNode>(ClassDB::instantiate(add_options[i].type));
```

Reply:
Currently this emit errors:

```
[ 89%] editor\plugins\visual_shader_editor_plugin.cpp(1235): error C2593: неоднозначный "operator ="
C:\Git\godot\core/object/ref_counted.h(144): note: может быть "void Ref<VisualShaderNode>::operator =(const Variant &)"
C:\Git\godot\core/object/ref_counted.h(127): note: или       "void Ref<VisualShaderNode>::operator =(const Ref<VisualShaderNode> &)"
editor\plugins\visual_shader_editor_plugin.cpp(1235): note: при попытке сопоставить список аргументов "(Ref<VisualShaderNode>, T *)"
        with
        [
            T=VisualShaderNode
        ]
```
Sorry, for behaving a Russian text, but it's the actual output of my compiler.

Reply:
Alright, then I'm proven wrong :D